job-runner/chain/reaper.go

```diff
@@ -0,0 +1,630 @@
+
+}
+
+// -------------------------------------------------------------------------- //
+
+// reaper is embedded in each type of JobReaper - it has fields and methods
+// they all use.
+type reaper struct {
+	chain             *Chain
+	chainRepo         Repo
+	rmc               rm.Client
+	logger            *log.Entry
+	finalizeTries     int
+	finalizeRetryWait time.Duration
+	doneJobChan       chan proto.Job
+	stopMux           *sync.Mutex
+	stopped           bool
+	stopChan          chan struct{}
+	doneChan          chan struct{}
+}
+
+// Sends the final state of the chain to the Request Manager, retrying a few times
+// if sending fails. It returns true if the final state was successfully sent;
+// else false.
+func (r *reaper) sendFinalState() {
+	err := retry.Do(r.finalizeTries, r.finalizeRetryWait,
+		func() error {
+			return r.rmc.FinishRequest(r.chain.RequestId(), r.chain.State())
+		},
+		nil,
+	)
+	if err != nil {
+		r.logger.Errorf("problem sending final status of the finished chain to the Request Manager: %s", err)
+	}
+}
+
+// prepareSequenceRetry prepares a sequence to be retried:
+//   1. Mark failed job and all previously completed jobs in sequence as PENDING.
+//   2. Decrement the number of jobs that were previously run (failed job +
+//   previously completed jobs in sequence) from the chain job count.
+//   3. Increment retry count for the sequence.
+func (r *reaper) prepareSequenceRetry(failedJob proto.Job) {
+	sequenceStartJob := r.chain.SequenceStartJob(failedJob.Id)
+
+	// sequenceJobsToRetry is a list containing the failed job and all previously
+	// completed jobs in the sequence. For example, if job C of A -> B -> C -> D
+	// fails, then A and B are the previously completed jobs and C is the failed
+	// job. So, jobs A, B, and C will be added to sequenceJobsToRetry. D will not be
+	// added because it was never run.
+	var sequenceJobsToRetry []proto.Job
+	sequenceJobsToRetry = append(sequenceJobsToRetry, failedJob)
+	sequenceJobsCompleted := r.sequenceJobsCompleted(sequenceStartJob)
+	sequenceJobsToRetry = append(sequenceJobsToRetry, sequenceJobsCompleted...)
+
+	// Mark all jobs to retry to PENDING state
+	for _, job := range sequenceJobsToRetry {
+		r.chain.SetJobState(job.Id, proto.STATE_PENDING)
+	}
+	r.chainRepo.Set(r.chain)
+}
+
+// sequenceJobsCompleted does a BFS to find all jobs in the sequence that have
+// completed. You can read how BFS works here:
+// https://en.wikipedia.org/wiki/Breadth-first_search.
+func (r *reaper) sequenceJobsCompleted(sequenceStartJob proto.Job) []proto.Job {
+	// toVisit is a map of job id->job to visit
+	toVisit := map[string]proto.Job{}
+	// visited is a map of job id->job visited
+	visited := map[string]proto.Job{}
+
+	// Process sequenceStartJob
+	for _, pJob := range r.chain.NextJobs(sequenceStartJob.Id) {
+		toVisit[pJob.Id] = pJob
+	}
+	visited[sequenceStartJob.Id] = sequenceStartJob
+
+PROCESS_TO_VISIT_LIST:
+	for len(toVisit) > 0 {
+
+	PROCESS_CURRENT_JOB:
```

Comment:
Maybe I'm missing something here, but once we get to the last job in the sequence, what's stopping this from continuing to look at that job's next jobs?

Reply:
Ack, thanks for tracking that down. Agreed, let's fix it in a future PR.